### ISSUE TITLE
EqualConstraint does not display the differences between JObject instances

### ISSUE THREAD

**Issue by reureu** [nunit/nunit#3949]
Hi there,

I am currently writing a test that compares 2 different JObjects.

The following test fails as expected.

```
using System;
using System.Collections.Generic;

using Newtonsoft.Json.Linq;

using NUnit.Framework;

namespace NUnitJObjectTests
{
    [TestFixture]
    public class JObjectTests
    {
        [Test]
        public void TestWithUnequalJObjects()
        {
            var expected = JObject.Parse(@"{ ""1"": 1, ""2"": { ""2.1"": 2.1 } }");
            var actual = JObject.Parse(@"{ ""1"": 1, ""2"": { ""2.1"": 2.11 } }");

            Assert.That(actual, Is.EqualTo(expected).Using(JToken.EqualityComparer));
        }
    }
}
```
But all it displays is:

`Expected and actual are both <Newtonsoft.Json.Linq.JObject> with 2 elements.`

I would expect NUnit to call JObjects.ToString() and display the result:

```
  Expected: <{
  "1": 1,
  "2": {
    "2.1": 2.1
  }
}>
  But was:  <{
  "1": 1,
  "2": {
    "2.1": 2.11
  }
}>
```

Here is what I tried:

I tried to add a ValueFormatter, but it doesn't get called.
`TestContext.AddFormatter<JObject>(j => j.ToString());`

So, I tried to dig into the NUnit.Framework code. Here is what I found out:
* EqualConstraintResult.DisplayDifferences sees that JObject implements ICollection and tries to display the differences by calling DisplayCollectionDifferences.
* EqualConstraintResult.DisplayCollectionDifferences does not have any failurePoints because the comparizon used the external comparer JToken.EqualityComparer.
* Consequently, no difference is written to the output.

Bearing that in mind, I changed
`Assert.That(actual, Is.EqualTo(expected).Using(JToken.EqualityComparer));`
for
`Assert.That(actual, Is.EquivalentTo(expected).Using(JToken.EqualityComparer));`
and I got something closer to the output I expected:
```
  Expected: equivalent to {
  "1": 1,
  "2": {
    "2.1": 2.1
  }
}
  But was:  {
  "1": 1,
  "2": {
    "2.1": 2.11
  }
}
  Missing (1): < < {
  "2.1": 2.1
} > >
  Extra (1): < < {
  "2.1": 2.11
} > >
```

By definition EquivalentTo interpretes the test objects as collections, but I need to test for equality.

So here are my questions;
* Why is the EqualConstraint unable to output the differences between actual and expected?
* Can an external comparer not be used to find out where the differences are when using an EqualConstraint?
* Alternatively, is there a way to call a custom ValueFormatter in such a case?

As a side note, I went for the following workaround
```
TestContext.AddFormatter<JObject>(j => j.ToString());
...
Assert.That(new[] { actual }, Has.One.Items.EqualTo(expected).Using(JToken.EqualityComparer));
```
which produced the following output
```
  Expected: exactly one item equal to {
  "1": 1,
  "2": {
    "2.1": 2.1
  }
}
  But was:  0 items < {
  "1": 1,
  "2": {
    "2.1": 2.11
  }
} >
```
That's not what I expected, but it's quite close.

**Comment by rprouse** [comment]
We are not going to provide comparisons between objects that are not in the core CLR libraries. Doing so would require us to take a dependency on the library in question and would be a never ending uphill battle. That is why we allow you to pass in custom comparers.

**Comment by reureu** [comment]
Hi @rprouse,

I understand your point, but the problem is not about comparing 2 objects. It's perfectly sensible to provide a custom comparer, and it is what I did. As I said, the test fails as expected.

**The problem is what nunit logs** if I just compare two different JObject's.

```
namespace NUnitJObjectTests
{
    [TestFixture]
    public class JObjectTests
    {
        [SetUp]
        public void Init()
        {
            TestContext.AddFormatter<JObject>(j => j.ToString());
        }

        [Test]
        public void TestWithUnequalJObjects()
        {
            var expected = JObject.Parse(@"{ ""1"": 1, ""2"": { ""2.1"": 2.1 } }");
            var actual = JObject.Parse(@"{ ""1"": 1, ""2"": { ""2.1"": 2.11 } }");

            Assert.That(actual, Is.EqualTo(expected).Using(JToken.EqualityComparer));
        }
    }
}
```
leads to the following output:
`Expected and actual are both <Newtonsoft.Json.Linq.JObject> with 2 elements.`

* Why is my custom formatter not called to display the content of `expected` and `actual`?
* Why is the custom formatter called if I reformulate my assertion like:
```Assert.That(new[] { actual }, Has.One.Items.EqualTo(expected).Using(JToken.EqualityComparer));```

Kind regards,
Reureu

**Comment by jnm2** [comment]
I agree that this is worth thinking about. If there is a formatter directly registered for a type, for instance `List<int>`, then we should use that formatter rather than our own collection presentation logic. (What else could have been intended?)

**Comment by stevenaw** [comment]
I believe this affects BCL types too, which I don't know if we want to require custom formatters for. Compare the following failure messages for `string[]` and `ImmutableArray<string>`
```csharp
[Test]
        public static void Compare_ImmutableArray()
        {
            var a = ImmutableArray.Create("a");
            var b = ImmutableArray.Create("b");

            Assert.That(a, Is.EqualTo(b));
            // Expected and actual are both <System.Collections.Immutable.ImmutableArray`1[System.String]> with 1 elements
        }

        [Test]
        public static void Compare_Array()
        {
            var a = new[] { "a" };
            var b = new[] { "b" };

            Assert.That(a, Is.EqualTo(b));
            // Expected and actual are both <System.String[1]>
            // Values differ at index[0]
            // String lengths are both 1.Strings differ at index 0.
            // Expected: "b"
            // But was:  "a"
            // ----------- ^
        }
```
